staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -309,6 +329,36 @@ func (w *watchCache) updateCache(event *watchCacheEvent) {
 	w.endIndex++
 }
 
+// Use dynamic size cache. If all cached events occur inside eventFreshDuration, increase cache capacity by 2x.
+// Decrease cache capacity by 2x when recent half events occur outside eventFreshDuration.
+func (w *watchCache) resizeCache(eventTime time.Time) {
+	if w.endIndex == w.startIndex+w.capacity && eventTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime) < eventFreshDuration {
```

Comment:
I'm probably paranoid here, but if our upperBound is smaller than what we would like to have, we will be evaluating this on every watch event add. Would you be able to add a simple benchmark that will propagate cache and then b.N times will try to `updateCache` (with events with consecutive RecordTime). And compare the results before and after this change?

Reply:
Yes, you are right. This evaluation causes double waste of time. Result measured by `BenchmarkWatchCache_updateCache`:
after this change:
BenchmarkWatchCache_updateCache-4       45565723                25.5 ns/op             0 B/op          0 allocs/op
before this change(without executing `resizeCache`):
BenchmarkWatchCache_updateCache-4       100000000               11.4 ns/op             0 B/op          0 allocs/op

Reply:
This is significant increase, but it's probably fine - these numbers are pretty low.
Note that in general for new watch cache event, we also do a bunch of things earlier and later (e.g. dispatch this event, process the incoming one etc.)

So the benchmark is great to have, but I guess it's fine...

Reply:
Thanks for your explain.